src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -574,7 +574,7 @@ private void MakeConstantTuple(LocalSymbol inProgress, BoundExpression boundInit
 
             internal override ConstantValue GetConstantValue(SyntaxNode node, LocalSymbol inProgress, BindingDiagnosticBag diagnostics = null)
             {
-                if (this.IsConst && inProgress == this)
+                if (this.IsConst && (inProgress == this || this.ForbiddenZone?.Contains(node) == true))
```

Comment:
> || this.ForbiddenZone?.Contains(node) == true) [](http://example.com/codeflow?start=55&length=47)

While, perhaps, this change prevents the stack overflow for the specific scenario, I think we should be addressing the root cause of the problem, i.e. the binder hierarchy used to perform binding. 
I am fine if we proceed with this change as a quick, low risk,  temporary fix. But we should follow-up on the issue around binder hierarchy.